### PR TITLE
Fixes some issues in the FortCL implementation

### LINKED ISSUE #11: Reading source files for JIT compilation sometimes fails with an encoding message
Sometimes (not always), some OpenCL devices fail when reading the OpenCL kernels source file for JIT compiling with the error:

1:168:1: error: source file is not valid UTF-8
<97>
^

This is fixed by adding a extra comment or line-break in the OpenCL kernels file, but this probably means there is an issue in the open()->allocate buffer->read() sequence in the get_program() subroutine

### LINKED ISSUE #12: Clean up compiler warnings
When compiling FortCL there is a very large number of compiler warnings, most of them come from the fact that the clfortran c_iso_bindings are functions instead of function interfaces.

### PR DESCRIPTION
**Pull request by sergisiso**
Fix #11 by reading the file in binary format and passing it as it is to the OpenCL runtime (avoids dealing with encoding)
Fix #12 by transforming OpenCL binded functions to function interfaces.

This implementation was branched from PR #10 branch, so that PR needs to go first.

### CODE REVIEW

**Comment by sergisiso** [comment]
@arporter This is ready for review now. This cleans up a big part of PSycloneBench makefile compiler warnings and fixes some encoding problems that appear from time to time when using the NemoLite2D FortCL and PSyclone generated OpenCL implementations.

File: src/clfortran.f90 [hunk 1]
```diff
@@ -1897,35 +2112,43 @@ integer(c_int32_t) function clEnqueueBarrierWithWaitList(command_queue, &
...
             type(c_ptr), value :: func_name
             
         end function
+    end interface
             
     !
     !end interface
     !
 
+    contains
```
Comment (arporter) [inline src/clfortran.f90]: I don't think you need this line (but I may be wrong)?

File: src/ocl_utils_mod.f90 [hunk 2]
```diff
@@ -160,53 +159,46 @@ function get_program(context, device, version_str, filename) result(prog)
...
-    allocate(source(irec+1), stat=iallocerr)
-    if (iallocerr.ne.0) stop 'allocate'
-    do i=1,irec
-       read(iunit,rec=i,iostat=ierr) source(i:i)
-    enddo
+
+    ! Get the size in bytes and allocate a byte buffer
+    inquire(iunit, SIZE=binary_size)
+    allocate(buffer(binary_size), stat=iallocerr)
+    if (iallocerr.ne.0) stop 'Error allocating buffer'
```
Comment (arporter) [inline src/ocl_utils_mod.f90]: Please extend a little to say which kernel/file and how many bytes we were attempting to allocate.

File: src/ocl_utils_mod.f90 [hunk 3]
```diff
@@ -160,53 +159,46 @@ function get_program(context, device, version_str, filename) result(prog)
...
-    end do
-    if (irec.eq.0) stop 'nothing read'
-    allocate(source(irec+1), stat=iallocerr)
-    if (iallocerr.ne.0) stop 'allocate'
-    do i=1,irec
-       read(iunit,rec=i,iostat=ierr) source(i:i)
-    enddo
+
+    ! Get the size in bytes and allocate a byte buffer
+    inquire(iunit, SIZE=binary_size)
```
Comment (arporter) [inline src/ocl_utils_mod.f90:177]: That's a lot nicer than the code I had!

File: src/clfortran.f90 [hunk 4]
```diff
@@ -426,39 +426,16 @@ module clfortran
...
     ! Start interfaces.
     !
-    contains
 
     ! ------------
     ! Platform API
     ! ------------
 
     ! clGetPlatformIDs
+    interface
```
Comment (arporter) [inline src/clfortran.f90:438]: I think this lack of the use of `interface` must be the reason I had terrible trouble originally when I included clfortran.o in the objects being compiled - I got a binary but it did nothing.

**Review by arporter — CHANGES_REQUESTED** [review]
Looks good and a big improvement over what we had originally.
I can build the lib on my desktop (although I can't run the examples because I no longer have a working OpenCL installation).
Only issue of any significance is to improve the verbosity of an error message. Once that's done I will merge.

File: src/clfortran.f90 (hunk 1 above)
Comment (sergisiso) [reply src/clfortran.f90]: Yeah, it works without it.

File: src/ocl_utils_mod.f90 (hunk 2 above)
Comment (sergisiso) [reply src/ocl_utils_mod.f90]: Improved error message

**Comment by sergisiso** [comment]
@arporter I made the suggested changes, ready for next review.

**Review by arporter — APPROVED** [review]

**Pull request merged**